conans/test/functional/test_third_party_patch_flow.py

```diff
@@ -0,0 +1,199 @@
+import os
+import textwrap
+
+from conans.test.utils.scm import create_local_git_repo
+from conans.test.utils.test_files import temp_folder
+from conans.test.utils.tools import TestClient
+from conans.util.files import mkdir
+
+
+def test_third_party_patch_flow():
+    """ this test emulates the work of a developer contributing recipes to ConanCenter, and having
+    to do multiple patches to the original library source code:
+    - Everything is local, not messing with the cache
+    - Using layout() to define location of things
+    """
+    conanfile = textwrap.dedent(r"""
+        import os
+        from conans import ConanFile, load
+        from conans.tools import save
+        from conan.tools.files import apply_conandata_patches
+
+        class Pkg(ConanFile):
+            name = "mypkg"
+            version = "1.0"
+            exports_sources = "*"
+
+            def layout(self):
+                self.folders.source = "src"
+                self.folders.build = "build"
+
+            def source(self):
+                # emulate a download from web site
+                save("myfile.cpp", "mistake1\nsomething\nmistake2\nmistake3\nsome\n")
+                apply_conandata_patches(self)
+
+            def build(self):
+                for i in (1, 2, 3):
+                    if "mistake{}".format(i) in load(os.path.join(self.source_folder, "myfile.cpp")):
+                        raise Exception("MISTAKE{} BUILD!".format(i))
+        """)
+
+    client = TestClient()
+    client.save({"conanfile.py": conanfile,
+                 "conandata.yml": ""})
+    client.run("install .")
+    client.run("source .")
+    client.run("build .", assert_error=True)
+    assert "MISTAKE1 BUILD!" in client.out
+
+    # user decides to create patches, first init the repo
+    client.init_git_repo(folder="src")  # Using helper for user/email repo init
+    client.save({"src/myfile.cpp": "correct1\nsomething\nmistake2\nmistake3\nsome\n"})
+    # compute the patch
+    mkdir(os.path.join(client.current_folder, "patches"))
+    client.run_command("cd src && git diff > ../patches/patch1")
+    client.run_command("cd src && git add . && git commit -m patch1")
+    conandata = textwrap.dedent("""
+        patches:
+          "1.0":
+            - patch_file: "patches/patch1"
+        """)
+    client.save({"conandata.yml": conandata})
+
+    client.run("source .")
```

Comment:
I would love to see the check it (`source` command) doesn't lose any of my local modifications made into the `src` sub-directory, if any.
right now it saves entire changes to the patch, it would be better if the test only saved some changes and some were just left as uncommitted or staged.